### PR TITLE
Optional Argument to dev_setup.py - Codifying Dev Dependencies

### PR DESCRIPTION
**Pull request by scbedd**
Context:

* We want to be able to build only what's changing/is changed/what we target
* dev_setup.py is a good entry point to automatically install packages for tests
* dev_setup.py only sets up ALL packages
* dev dependencies (aka only required by tests which are unpackaged) are not codified anywhere currently

I did not go quite as simple as you mentioned @lmazuel. I actually expand all dependencies recursively, this is to ensure that we catch dev dependencies of `azure-sdk-tools` in addition to whatever package is being targeted. 

Behavior when no argument is provided is identical to current usage.

In addition, note that `azure-sdk-tools` is *explicitly* called out in many of the `dev_requirements` files. I could absolutely introduce this as an "assumed" dependency, and reduce the usage of `dev_requirements` in most cases. However, I lean towards being very clear that the dependency is actually present. This choice can be amended (and a bunch of dev_requirements thrown out) depending on what owners/maintainers think.

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: azure-mgmt-subscription/dev_requirements.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+azure-sdk-tools
+azure-mgmt-billing
```
Comment (lmazuel) [inline azure-mgmt-subscription/dev_requirements.txt]: sounds like a mistake, I don't see how billing could be a dependency here, maybe a failed copy/paste in the test

File: azure-mgmt-batch/dev_requirements.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+azure-sdk-tools
+azure-common
```
Comment (lmazuel) [inline azure-mgmt-batch/dev_requirements.txt]: Don't need common in dev_dependencies, it's a runtime dependency already

File: azure-mgmt-eventgrid/dev_requirements.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+azure-sdk-tools
+azure-common
```
Comment (lmazuel) [inline azure-mgmt-eventgrid/dev_requirements.txt]: Don't need common in dev_dependencies, it's a runtime dependency already

File: azure-mgmt-eventhub/dev_requirements.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+azure-sdk-tools
+azure-common
```
Comment (lmazuel) [inline azure-mgmt-eventhub/dev_requirements.txt]: Don't need common in dev_dependencies, it's a runtime dependency already

File: azure-mgmt-recoveryservices/dev_requirements.txt [hunk 5]
```diff
@@ -0,0 +1,3 @@
+azure-sdk-tools
+azure-common
```
Comment (lmazuel) [inline azure-mgmt-recoveryservices/dev_requirements.txt]: Don't need common in dev_dependencies, it's a runtime dependency already

File: azure-mgmt-relay/dev_requirements.txt [hunk 6]
```diff
@@ -0,0 +1,2 @@
+azure-sdk-tools
+azure-common
```
Comment (lmazuel) [inline azure-mgmt-relay/dev_requirements.txt]: Don't need common in dev_dependencies, it's a runtime dependency already

File: azure-mgmt-resource/dev_requirements.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+azure-sdk-tools
+azure-common
```
Comment (lmazuel) [inline azure-mgmt-resource/dev_requirements.txt]: Don't need common in dev_dependencies, it's a runtime dependency already

File: azure-mgmt-servicebus/dev_requirements.txt [hunk 8]
```diff
@@ -0,0 +1,2 @@
+azure-sdk-tools
+azure-common
```
Comment (lmazuel) [inline azure-mgmt-servicebus/dev_requirements.txt]: Don't need common in dev_dependencies, it's a runtime dependency already

File: azure-servicebus/dev_requirements.txt [hunk 9]
```diff
@@ -0,0 +1,2 @@
+azure-sdk-tools
+azure-common
```
Comment (lmazuel) [inline azure-servicebus/dev_requirements.txt]: Don't need common in dev_dependencies, it's a runtime dependency already

File: azure-servicemanagement-legacy/dev_requirements.txt [hunk 10]
```diff
@@ -0,0 +1,2 @@
+azure-sdk-tools
+azure-common
```
Comment (lmazuel) [inline azure-servicemanagement-legacy/dev_requirements.txt]: Don't need common in dev_dependencies, it's a runtime dependency already

**Comment by lmazuel** [comment]
I understand what you need "azure-common" installed, and if no "dev_requirements.txt" mention it, filtering to "azure-keyvault" will not find it. I would prefer to have a list of "always install these packages" more than playing with dev_requirements.txt (which is an official format).

The other question is "azure-sdk-tools". Technically, it's not possible to do "pip install -r dev_requirements.txt", because "azure-sdk-tools" is not on PyPI. On the other hand, writing "-e ../azure-sdk-tools" breaks the logic a little. So I think it's not perfect, but fair to keep "azure-sdk-tools"

**Comment by lmazuel** [comment]
@scbedd "azure-sdk-tools" being a dev package only, the distinction between runtime and dev dependencies is fuzzy for this guy.

I think the right way to fix this one is to update the "setup.py" of "azure-sdk-tools" to install in dev mode the two mentionned packages. Here:
https://github.com/Azure/azure-sdk-for-python/blob/master/azure-sdk-tools/setup.py

Let me think about it, it needs some experiments here.

**Comment by lmazuel** [comment]
@scbedd This PR is not critical for your release process, it will just make it faster, so I don't want to rush until we find the correct way to do it in the best way possible.

**Comment by lmazuel** [comment]
@scbedd I've been thinking about this a lot, this is my current take on it:
- Hard code "pip install -e azure-common" in dev_setup.py as step 1 (whatever the input filter)
- Hard code "pip install -e azure-sdk-tools" in dev_setup.py as step 2 (whatever the input filter)
- Change all dev_requirements.txt to do "-e ../azure-sdk-tools". Only this guy, everything else package name like you did (e.g. "azure-mgmt-authorization")
- Change the install part in the loop to do a two steps:
  - Step 1: pip install -e \<package name\> (with your recursive detection)
  - Step 2: pip install -r \<package name\>/dev_requirements.txt
- Add a dev dependency in editable mode in "azure-sdk-tools/setup.py" for authorization/resource. Exact syntax, I need to double check setuptools, editable dependency is a rare syntax for a setup.py (but doable)
- Move these 4 lines into the proper dev_requirements.txt (as mentionned in the comment)

https://github.com/Azure/azure-sdk-for-python/blob/26611f2b2ed1e67c120884e2bccefefbae94671a/azure-sdk-tools/setup.py#L22-L25

Then we should have something that:
- Make a small dev_setup if necessary with input filter
- Works as expected for a dev_requirements.txt (can be autonomous for dev people)

**Pull request merged**